### PR TITLE
feat: add `carousel` component

### LINKED ISSUE #226: [Feature]: Carousel component (Embala)
### Describe the feature

New Component

[Vue Embla Carousel](https://www.embla-carousel.com/get-started/vue/)

From #220 

### Additional information

- [X] I intend to submit a PR for this feature.
- [X] I have already implemented and/or tested this feature.

### PR DESCRIPTION
**Pull request by wasimTQ**
- Added caroused for default styles
- Added example for orientation vertical

fixes #226 

I need help on how to add it to the table of contents

Still there are some examples that needs to be done. And also it has to be copied to new york styling as well

### CODE REVIEW

**Comment by wasimTQ** [comment]
@sadeghbarati I stuck on how we can include the page in the table of contents. Can you help me out with it.

Also how can I do something like this `apps/www/src/public/registry/styles/default/calendar.json`

**Comment by sadeghbarati** [comment]
Thanks buddy will check it

**Comment by wasimTQ** [comment]
@sadeghbarati Pretty much everything is done. I just have to know how I can generate the `.json` files.

Can you explain how can I do that? 

Also please let me know if I did it all correct. Thank you

**Comment by sadeghbarati** [comment]
https://github.com/radix-vue/shadcn-vue/assets/17789047/74080157-2f2d-4c8b-bd4a-c0a06a07b638

**Comment by wasimTQ** [comment]
@sadeghbarati Hey man, sorry I can't replicate the issue. I'll see what I can do

**Comment by sadeghbarati** [comment]
I can't replicate the issue either with another Windows system, guess it was my system issue then

File: apps/www/src/lib/registry/default/example/CarouselPlugin.vue
Comment (sadeghbarati) [inline apps/www/src/lib/registry/default/example/CarouselPlugin.vue]: Why `ref`, is their Autoplay plugin is reactive when we change the options?

**Comment by wasimTQ** [comment]
Also I'm looking to make the carousel a focusable element for easier keyboard interactions.

I've looked in the internet. `tabindex="0"` is the answer I got. But the problem is it don't unfocus on tab click once it's been focused

**Comment by sadeghbarati** [comment]
We should change `ComponentPreview` tabindex when it's about Carousel demos

For focusable carousel just add tabindex 0 to `Carousel.vue`

File: apps/www/src/lib/registry/default/ui/carousel/Carousel.vue
Comment (sadeghbarati) [inline apps/www/src/lib/registry/default/ui/carousel/Carousel.vue]: This is pointless when you forwarded the props and emits on the native `div` element can you remove it in both styles?

image

image


`useProvideCarousel` will handle props and emits right?

File: apps/www/src/lib/registry/default/example/CarouselPlugin.vue
Comment (wasimTQ) [reply apps/www/src/lib/registry/default/example/CarouselPlugin.vue]: No. It's just to be ensure that the autoplay plugin initialize only once.

Comment (wasimTQ) [reply apps/www/src/lib/registry/default/example/CarouselPlugin.vue]: Maybe I should change the ref to reactive or so

File: apps/www/src/lib/registry/default/ui/carousel/Carousel.vue
Comment (wasimTQ) [reply apps/www/src/lib/registry/default/ui/carousel/Carousel.vue]: I removed it

Comment (wasimTQ) [reply apps/www/src/lib/registry/default/ui/carousel/Carousel.vue]: You can review it once. I still have some documentation to add. I'll do that now

**Comment by sadeghbarati** [comment]
For building registry you can run `build:registry` in `apps/www` directory, it's optional but you can ignore it and let Zernonia build it before the release

**Comment by wasimTQ** [comment]
@sadeghbarati I think I'm pretty much done. Can you please review it once more and let me know when you're free?

File: apps/www/src/public/registry/index.json
Comment (sadeghbarati) [inline apps/www/src/public/registry/index.json]: Can you include missing file `useCarousel.ts` and move `interface.ts` content inside `index.ts`?

Comment (wasimTQ) [reply apps/www/src/public/registry/index.json]: Ok. I'll do it.

**Comment by sadeghbarati** [comment]
Only one thing which I doubt to include it in `Carousel.vue` or not

```ts
defineExpose({
  carouselRef: carouselArgs or carouselApi
})
```

So the user can access the Carousel API in every possible way in Vue

- [Ref on Component](https://vuejs.org/guide/essentials/template-refs#ref-on-component)
- `@init-api` emit
- `useCarousel` access Carousel API in component's descendants
- slotProps

File: apps/www/src/public/registry/index.json
Comment (wasimTQ) [reply apps/www/src/public/registry/index.json]: Why moving `interface.ts` content to `index.ts` tho?. 

I'd say this'll be best for readability

**Comment by wasimTQ** [comment]
It's already accessible through slotProps. I'll check if setting defineExpose is important.

I think it's a nice to have. I'll add it

**Comment by wasimTQ** [comment]
@sadeghbarati I've made all the changes. You can check it when you're free and let me know

**Review by sadeghbarati — APPROVED** [review]

**Comment by sadeghbarati** [comment]
https://github.com/radix-vue/shadcn-vue/pull/227#issuecomment-1872250822

Sometimes I still have this issue but it will be resolved after embla `8.0.0-rc18`, Thanks to the `embla-carousel` author
Let's wait until that version gets published after that we can merge this

**Comment by zernonia** [comment]
> [#227 (comment)](https://github.com/radix-vue/shadcn-vue/pull/227#issuecomment-1872250822)
> 
> Sometimes I still have this issue but it will be resolved after embla `8.0.0-rc18`, Thanks to the `embla-carousel` author Let's wait until that version gets published after that we can merge this

@sadeghbarati  do you have the relevant ticket to the issue you mentioned? 😁

**Comment by sadeghbarati** [comment]
Hey, Yes here 

https://github.com/davidjerleke/embla-carousel/issues/672

**Comment by zernonia** [comment]
Thanks @sadeghbarati ..


Something weird in the `build:registry`, I can't seems to run them lol

```
/Users/zernonia/Desktop/RadixVue/shadcn-vue/node_modules/.pnpm/@vue+compiler-sfc@3.3.7/node_modules/@vue/compiler-sfc/dist/compiler-sfc.cjs.js:15656
    throw new Error(
          ^

Error: [@vue/compiler-sfc] Failed to resolve import source "./interface".

anonymous.vue
4  |  import { cn } from '@/lib/utils'
5  |  
6  |  const props = withDefaults(defineProps<CarouselProps>(), {
   |                                         ^^^^^^^^^^^^^
7  |    orientation: 'horizontal',
8  |  })
```

**Comment by sadeghbarati** [comment]
Sadly Vue doesn't support Complex Types (entire props object) in SFC file, I thought I would solve this issue by defining `opts` objects key types one by one like, but it's still not working  

```ts
export interface CarouselProps {
  opts?: {
    active: CarouselOptions['active']
    align: CarouselOptions['align']
    axis: CarouselOptions['axis']
    breakpoints: CarouselOptions['breakpoints']
    container: CarouselOptions['container']
    containScroll: CarouselOptions['containScroll']
    direction: CarouselOptions['direction']
    dragFree: CarouselOptions['dragFree']
    dragThreshold: CarouselOptions['dragThreshold']
    duration: CarouselOptions['duration']
    inViewThreshold: CarouselOptions['inViewThreshold']
    loop: CarouselOptions['loop']
    skipSnaps: CarouselOptions['skipSnaps']
    slides: CarouselOptions['slides']
    slidesToScroll: CarouselOptions['slidesToScroll']
    startIndex: CarouselOptions['startIndex']
    watchDrag: CarouselOptions['watchDrag']
    watchResize: CarouselOptions['watchResize']
    watchSlides: CarouselOptions['watchSlides']
  }
  ...
}
```


cause `EmblaOptionsType` (OptionsType) is using `CreateOptionsType` which is using `extends`


#### `CarouselOptions types definitions`

```ts
export type EmblaOptionsType = Partial<OptionsType>;

// here is where trouble started, this types is complex
export type OptionsType = CreateOptionsType<{...}> 

// OptionType also extends other type
export type CreateOptionsType<Type extends LooseOptionsType> = Type & {
    active: boolean;
    breakpoints: {
        [key: string]: Omit<Partial<CreateOptionsType<Type>>, 'breakpoints'>;
    };
};
```


---

Hands down to Vue Reactivity system, it is best, but their TypeScript support is in mid level 😶‍🌫️

**Comment by zernonia** [comment]
Yeah. I wish the type system for Vue would be improved soon! Maybe lets try this? https://github.com/so1ve/vue.ts/tree/main/packages/complex-types

**Comment by sadeghbarati** [comment]
I tried to exclude carousel files from `vue-tsc` typecheck but vue-tsc ignored exclude property in `tsconfig.registry.json`

https://github.com/vuejs/language-tools/issues/3144


```json
{
  "extends": "@vue/tsconfig/tsconfig.json",
  "compilerOptions": {
    "moduleResolution": "Node",
    "baseUrl": ".",
    "paths": {
      "@/*": ["./src/*"]
    },
    "declaration": false
  },
  "include": ["src/lib/**/*"],
  "exclude": ["node_modules", "src/lib/registry/**/example/**/*", "src/lib/registry/default/ui/carousel/*", "src/lib/registry/new-york/ui/carousel/*"]
}
```

---

in this case we should build registry without vue-tsc typecheck

**Comment by sadeghbarati** [comment]
@wasimTQ can you update `embla-carousel-vue` to latest version, thanks

**Comment by wasimTQ** [comment]
@sadeghbarati It's done

**Comment by sadeghbarati** [comment]
@wasimTQ 

Thanks, just one more thing before other maintainer merges it into the main 

can you refactor the `$attrs.class ?? ''` to `props.class`

Vue `inheritAttrs` Component Option is **true** by default so `$attrs.class ?? ''` will cause duplicate class but with `defineProps` we **omit** class from `$attrs` object so we can prevent class duplication

I'll share example


<details><summary>Carousel.vue</summary>
<p>

```vue
<script setup lang="ts">
import type { HTMLAttributes } from 'vue'
import { useProvideCarousel } from './useCarousel'
import type { CarouselEmits, CarouselProps } from './interface'
import { cn } from '@/lib/utils'

const props = withDefaults(defineProps<CarouselProps & { class?: HTMLAttributes['class'] }>(), {
  orientation: 'horizontal',
})
const emits = defineEmits<CarouselEmits>()
const carouselArgs = useProvideCarousel(props, emits)

function onKeyDown(event: KeyboardEvent) {
  const prevKey = props.orientation === 'vertical' ? 'ArrowUp' : 'ArrowLeft'
  const nextKey = props.orientation === 'vertical' ? 'ArrowDown' : 'ArrowRight'
  if (event.key === prevKey) {
    event.preventDefault()
    carouselArgs.scrollPrev()
    return
  }
  if (event.key === nextKey) {
    event.preventDefault()
    carouselArgs.scrollNext()
  }
}

defineExpose(carouselArgs)
</script>

<template>
  <div
    :class="cn('relative', props.class)"
    role="region"
    aria-roledescription="carousel"
    tabindex="0"
    @keydown="onKeyDown"
  >
    <slot v-bind="carouselArgs" />
  </div>
</template>
```

</p>
</details> 


<details><summary>CarouselContent.vue</summary>
<p>

```vue
<script setup lang="ts">
import type { HTMLAttributes } from 'vue'
import { useCarousel } from './useCarousel'
import { cn } from '@/lib/utils'

defineOptions({
  inheritAttrs: false
})

const props = defineProps<{ class?: HTMLAttributes['class']  }>()

const { carouselRef, orientation } = useCarousel()
</script>

<template>
  <div ref="carouselRef" class="overflow-hidden">
    <div
      :class="
        cn(
          'flex',
          orientation === 'horizontal' ? '-ml-4' : '-mt-4 flex-col',
          props.class,
        )"
       v-bind="$attrs"
    >
      <slot />
    </div>
  </div>
</template>
```

</p>
</details> 

<details><summary>CarouselItem.vue</summary>
<p>

```vue
<script setup lang="ts">
import type { HTMLAttributes } from 'vue'
import { useCarousel } from './useCarousel'
import { cn } from '@/lib/utils'

const props = defineProps<{ class?: HTMLAttributes['class']  }>()

const { orientation } = useCarousel()
</script>

<template>
  <div
    role="group"
    aria-roledescription="slide"
    :class="cn(
      'min-w-0 shrink-0 grow-0 basis-full',
      orientation === 'horizontal' ? 'pl-4' : 'pt-4',
      props.class,
    )"
  >
    <slot />
  </div>
</template>
```

</p>
</details> 


<details><summary>CarouselNext.vue and CarouselPrevious.vue</summary>
<p>

```vue
<script setup lang="ts">
import type { HTMLAttributes } from 'vue'
import { ChevronRight } from 'lucide-vue-next'
import { useCarousel } from './useCarousel'
import { cn } from '@/lib/utils'
import { Button } from '@/lib/registry/default/ui/button'

const props = defineProps<{ class?: HTMLAttributes['class']  }>()

const { orientation, canScrollNext, scrollNext } = useCarousel()
</script>

<template>
  <Button
    :disabled="!canScrollNext"
    :class="cn(
      'absolute h-10 w-10 rounded-full p-0',
      orientation === 'horizontal'
        ? '-right-12 top-1/2 -translate-y-1/2'
        : '-bottom-12 left-1/2 -translate-x-1/2 rotate-90',
      props.class,
    )"
    variant="outline"
    @click="scrollNext"
  >
    <ChevronRight class="h-4 w-4 text-white stroke-white" />
  </Button>
</template>
```

</p>
</details>

**Comment by sadeghbarati** [comment]
Thanks, just one more thing before other maintainer merges it into the main 

can you refactor the `$attrs.class ?? ''` to `props.class`

Vue `inheritAttrs` Component Option is **true** by default so `$attrs.class ?? ''` will cause duplicate class but with `defineProps` we **omit** class from `$attrs` object so we can prevent class duplication

**Comment by Tajcore** [comment]
Curious while testing this branch I notice that `embla-carousel-vue` package doesn't explicitly exports the `EmblaCarouselType` `EmblaCarouselType` and `EmblaPluginType` type defs, and from this [post](https://github.com/davidjerleke/embla-carousel/issues/647#issuecomment-1868273239) the maintainer for the package mentioned it would be best to import those from the `embla-carousel` package instead of the Vue port.

**Comment by sadeghbarati** [comment]
@Tajcore 

https://github.com/davidjerleke/embla-carousel/issues/672#issuecomment-1872988293

**Comment by wasimTQ** [comment]
@sadeghbarati I used a different approach for solving this issue. Can you check it and let me know if you want it to be changes to the one you suggested.

**Comment by sadeghbarati** [comment]
@wasimTQ 

- `inhritAttrs: false` works just fine but just for the class, what if the user wants to add another attribute like `data-slot` or any other attributes?

- Demo arrow/chevron left/right is not visible in light mode, `text-current` will solve this issue

image


--- 

If you are busy, I can do it for you, so let me know 🙌

**Comment by wasimTQ** [comment]
Got it.

**Comment by wasimTQ** [comment]
@sadeghbarati I've made the changes.

**Comment by sadeghbarati** [comment]
@zernonia 

Problem is not complex types in fact if there was complex type error we can't even open carousel.md path in the localhost and should see vite error overlay on screen

 The problem is `@vue/compiler-sfc` can't resolve types in registry build, I remember "Dunqing" had this issue in `packages/cli` too which Idk how he solved it

 
 ---
 
 
 :x:
 
 ```ts
 import type { CarouselEmits, CarouselProps, WithClassAsProps } from './interface'
 
 const props = withDefaults(defineProps<CarouselProps & WithClassAsProps>(), {
  orientation: 'horizontal',
})
 ```
 
`[@vue/compiler-sfc] Failed to resolve import source "./interface"`
 
 
 ---
 
 ✅ 
 
 
  ```ts
  import type { HTMLAttributes, Ref } from 'vue'
 import type {
    EmblaCarouselType as CarouselApi,
    EmblaOptionsType as CarouselOptions,
    EmblaPluginType as CarouselPlugin,
} from 'embla-carousel'
 
 const props = withDefaults(defineProps<{
    opts?: CarouselOptions | Ref<CarouselOptions>
    plugins?: CarouselPlugin[] | Ref<CarouselPlugin[]>
    orientation?: 'horizontal' | 'vertical'
    class?: HTMLAttributes['class']
  }>(), {
    orientation: 'horizontal',
  })
 ```

**Comment by zernonia** [comment]
Thanks @sadeghbarati . Found the issue with `@vue/compiler-sfc` and fixed it. Now generating the registry correctly and as expected.

**Review by zernonia — APPROVED** [review]
Awesome job @wasimTQ  @sadeghbarati ! You 2 absolutely nailed it!! 
Thanks for the care for details, and the thorough implementation! 💚

LGTM and let's get this baby in!

**Pull request merged**
